core/field_matrix.js

```diff
@@ -0,0 +1,456 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2016 Massachusetts Institute of Technology
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview 5x5 matrix input field.
+ * Displays an editable 5x5 matrix for controlling LED arrays.
+ * @author khanning@gmail.com (Kreg Hanning)
+ */
+'use strict';
+
+goog.provide('Blockly.FieldMatrix');
+
+goog.require('Blockly.DropDownDiv');
+
+/**
+ * Class for a matrix field.
+ * @param {number} matrix The default matrix value represented by a 25-bit integer.
+ * @extends {Blockly.Field}
+ * @constructor
+ */
+Blockly.FieldMatrix = function(matrix) {
+  Blockly.FieldMatrix.superClass_.constructor.call(this, matrix);
+  this.addArgType('matrix');
+};
+goog.inherits(Blockly.FieldMatrix, Blockly.Field);
+
+/**
+ * Construct a FieldMatrix from a JSON arg object.
+ * @param {!Object} options A JSON object with options (matrix).
+ * @returns {!Blockly.FieldMatrix} The new field instance.
+ * @package
+ * @nocollapse
+ */
+Blockly.FieldMatrix.fromJson = function(options) {
+  return new Blockly.FieldMatrix(options['matrix']);
+};
+
+/**
+ * Fixed width of the matrix thumbnail in drop down menu, in px.
+ * @type {number}
+ * @const
+ */
+Blockly.FieldMatrix.MATRIX_THUMB_SIZE = 26;
+
+/**
+ * Fixed size of arrow icon in drop down menu, in px.
+ * @type {number}
+ * @const
+ */
+Blockly.FieldMatrix.ARROW_SIZE = 12;
+
+/**
+ * Fixed width of each button inside the 5x5 matrix, in px.
+ * @type {number}
+ * @const
+ */
+Blockly.FieldMatrix.MATRIX_NODE_SIZE = 18;
+
+/**
+ * Fixed corner radius for 5x5 matrix buttons, in px.
+ * @type {number}
+ * @const
+ */
+Blockly.FieldMatrix.MATRIX_NODE_RADIUS = 4;
+
+/**
+ * Fixed padding for 5x5 matrix buttons, in px.
+ * @type {number}
+ * @const
+ */
+Blockly.FieldMatrix.MATRIX_NODE_PAD = 5;
+
+/**
+ * Called when the field is placed on a block.
+ * @param {Block} block The owning block.
+ */
+Blockly.FieldMatrix.prototype.init = function() {
+  if (this.fieldGroup_) {
+    // Matrix menu has already been initialized once.
+    return;
+  }
+
+  // Build the DOM.
+  this.fieldGroup_ = Blockly.utils.createSvgElement('g', {}, null);
+  this.size_.width = Blockly.FieldMatrix.MATRIX_THUMB_SIZE +
+    Blockly.FieldMatrix.ARROW_SIZE + (Blockly.BlockSvg.DROPDOWN_ARROW_PADDING * 1.5);
+
+  this.sourceBlock_.getSvgRoot().appendChild(this.fieldGroup_);
+
+  this.getSourceColours_();
+
+  if (!this.ledThumbs_) {
```

Comment:
General comment:
Any properties that the field has should be explicitly declared and annotated, often in the constructor, unless they're declared in the superclass.

This isn't done in most of the existing fields because they got created before I understood how helpful these annotations are.

Benefits:
- When debugging down the line, you won't need to search for the first place that a property is used to know what it is
- Automated type checking can catch bugs (if it's being used)
- The next person editing this code won't come find where you live to understand it
- You catch places where you're mixing types in an unsafe way
- You can explicitly set the default value, rather than checking for undefined

Downsides:
- Constructors become gigantic, and most of it is in comments
- Takes time
- ???

Blockly's [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide#jsdoc) describes what annotations should look like